### PR TITLE
Feature: Show coverage area for existing stations and towns

### PR DESCRIPTION
**Pull request by PeterN**
This PR adds the following:

* Button in station window to toggle showing coverage area.
* Button in town window to toggle showing covered tiles.
* When placing a new station part, if the part would touch an existing station, that station's coverage area is drawn.

### CODE REVIEW

**Comment by andythenorth** [comment]
Station with coverage toggled off.  After closing the window, coverage is shown.


7446_close_bug_1
7446_close_bug_2

**Comment by PeterN** [comment]
Logic fail. Logic now tweaked. Logic better. (I hope.)

**Comment by andythenorth** [comment]
Caught a bug:
* find an existing rail station, but do not open the station window
* open the station construction tool and add one or more tiles directly to the existing station
* coverage will correctly be shown
* leave the mouse hovered over the station, then use esc to clear the construction tool
* coverage will continue to be shown, it should be cleared at this point

**Comment by andythenorth** [comment]
Works for me.  

There's a case where coverage turns off when station building tools are selected.  It's kind of unexpected, but it's consistent and not broken.

**Comment by PeterN** [comment]
Fixed up and rebased to master.

File: src/viewport.cpp [hunk 1]
```diff
@@ -979,6 +979,67 @@ static void DrawAutorailSelection(const TileInfo *ti, uint autorail_type)
...
 
+enum TileHighlightType {
+	THT_NONE,
+	THT_WHITE,
+	THT_BLUE,
+	THT_RED,
+};
+
+const Station *_station_th; ///< Currently selected station for coverage area highlight
+const Town *_town_th;       ///< Currently selected town for coverage area highlight
```
Comment (nielsmh) [inline src/viewport.cpp]: These names definitely made me "huh?" for a while until I discovered th = tile highlight. Can I suggest `_viewport_highlight_town` or similar?

File: src/viewport_func.h [hunk 2]
```diff
@@ -93,4 +93,10 @@ static inline void MarkTileDirtyByTile(TileIndex tile, int bridge_level_offset =
 
 Point GetViewportStationMiddle(const ViewPort *vp, const Station *st);
 
+struct Station;
+struct Town;
+
+void ViewportCatchmentStation(const Station *st, bool sel);
+void ViewportCatchmentTown(const Town *t, bool sel);
```
Comment (nielsmh) [inline src/viewport_func.h]: I think these need a Set or Show in the function name. Without a verb they sound like type names.

File: src/viewport.cpp (hunk 1 above)
Comment (PeterN) [reply src/viewport.cpp]: Done

File: src/viewport_func.h (hunk 2 above)
Comment (PeterN) [reply src/viewport_func.h]: Also done

**Comment by PeterN** [comment]
Updated so that town coverage (if shown) stays on when placing a station.

**Comment by nielsmh** [comment]
Two minor issues:
- Pushing Ctrl while building a station, and the station is in a would-join position does not remove the would-be-joined-to station's coverage immediately, but if you move the selection to a different tile and back while holding Ctrl it no longer shows the would-be-joined-to station's coverage. Same if you release Ctrl.
- Hovering over the "Build a separate station" option in the "Join station" window highlights the coverage of the first station in the list, instead of showing only the new station coverage.

**Review by nielsmh — DISMISSED** [review]

**Review by nielsmh — APPROVED** [review]

**Pull request merged**